### ISSUE TITLE
Loading individual Variables using EntryFilter API below the root level

### ISSUE THREAD

**Issue by MieszkoMakuch** [HebiRobotics/MFL#69]
Hi, I know that it is possible to load individual variables on the root level using the [EntryFilter](https://github.com/HebiRobotics/MFL/blob/9775216536ff07274bbdbb80693bb0d003e23022/mfl-core/src/main/java/us/hebi/matlab/mat/format/Mat5Reader.java#L114-L123) API. However, would it be possible to add an option to load individual variables "below" the root level? I want to read a file which has "**Catalog**" struct with multiple rows. I want to read each of the rows of this struct separately without changing the file structure. Here is an example file: [TG_catalog_205638_seismic_events (1) 4 mb.mat.zip](https://github.com/HebiRobotics/MFL/files/9226551/TG_catalog_205638_seismic_events.1.4.mb.mat.zip). I want to read e.g. ID, Time, Lat rows separately due to memory restrictions.

Here is the preview of the "**Catalog**" struct inside the file:
|           |     |                 |        |                                                                                  |             |
|-----------|-----|-----------------|--------|----------------------------------------------------------------------------------|-------------|
| 'ID'      | 3   | 205638×1 cell   | 'char' | 'Event ID'                                                                       | [ ]         |
| 'Time'    | 5   | 205638×1 double | [ ]    | 'Event origin time'                                                              | [ ]         |
| 'Lat'     | 25  | 205638×1 double | 'deg'  | 'Latitude'                                                                       | [ ]         |
| 'Long'    | 35  | 205638×1 double | 'deg'  | 'Longitude'                                                                      | [ ]         |
...
| 'RakeA'   | 130 | 205638×1 double | 'deg'  | 'Rake of nodal plane A'                                                          | [ ]         |

**Comment by ennerf** [comment]
The filter unfortunately only works on the root level as entries are compressed independently of each other and can be skipped. Structs are saved as a single variable and you always have to process the whole thing.

Can you store your struct as root level variables? The `save` command has flags for converting structs to and from root level entries.

```Matlab
myStruct = struct()
myStruct.a = 2
myStruct.b =3
save('test.mat', '-struct', 'myStruct')
myStruct  = load('test.mat')
```

That way you can filter `a` and `b`.

**Comment by MieszkoMakuch** [comment]
Thanks for the information. 

I also wanted to ask you if you have some suggestions on how to read such file in memory-efficient way. This particular file is quite small but we want to read much larger files (e.g. 10x larger).

My first step was to use the [EntryFilter](https://github.com/HebiRobotics/MFL/blob/9775216536ff07274bbdbb80693bb0d003e23022/mfl-core/src/main/java/us/hebi/matlab/mat/format/Mat5Reader.java#L114-L123) API.

Secondly, I tried the [Mat5Reader::setBufferAllocator](https://github.com/HebiRobotics/MFL/blob/9775216536ff07274bbdbb80693bb0d003e23022/mfl-core/src/main/java/us/hebi/matlab/mat/format/Mat5Reader.java#L142-L154) API to generate memory-mapped files. This works well for double/integer arrays but I see that the problem is the cell array of strings (the `'ID'` field in my example file). Is there any way to map such filed to memory mapped file?

**Comment by MieszkoMakuch** [comment]
I saw one of the previous issues #66 where you suggested to transform String cell to the 2D char array. I also tried this approach. I transformed a test file so that is has a 2D char array [TG_catalog_2DcharArray.mat.zip](https://github.com/HebiRobotics/MFL/files/9227906/TG_catalog_2DcharArray.mat.zip). 

However, I see that despite using [Mat5Reader::setBufferAllocator](https://github.com/HebiRobotics/MFL/blob/9775216536ff07274bbdbb80693bb0d003e23022/mfl-core/src/main/java/us/hebi/matlab/mat/format/Mat5Reader.java#L142-L154) the char array is not mapped to the file. In [CharEncoding#readCharBuffer](https://github.com/HebiRobotics/MFL/blob/43b24e8da07eed5c1ea5dca2c1ef8105d1cd2e3e/mfl-core/src/main/java/us/hebi/matlab/mat/format/CharEncoding.java#L92) for UTF the default buffer allocator is used. Additionally when converting to char buffer additional in memory buffer is allocated which increases memory usage even further.

**Comment by ennerf** [comment]
You are right that the Char encoding currently always uses heap buffers. I'll make a note to add support for custom allocators.

Just how memory constrained is your system? Even a 10x increase is only a 40MB file. Is the data heavily compressed and gets significantly larger? Is it a heap-size issue or a an issue with total system memory?

**Comment by MieszkoMakuch** [comment]
The data is quite heavily compressed. 4mb compressed file is around 30mb when uncompressed. Reading such file takes around 60mb of memory. As we want to support files that are 10x larger this means around 600mb of system memory per file.

We have a quite complex web system where reading mat files is just a small part. In addition to that multiple users may request multiple files at the same time. We established a memory constraint of 100 mb per file. That is why we would like to reduce the memory consumption as much as possible.

**Comment by ennerf** [comment]
Dealing with char buffers turned out to be really finicky, but they should now be properly using buffer allocators: [CharEncoding.java#L117-L182](https://github.com/HebiRobotics/MFL/blob/f605a73bfd16b1606ed64e1b4068fa0c242462d7/mfl-core/src/main/java/us/hebi/matlab/mat/format/CharEncoding.java#L117-L182)

I pushed `0.6.0-SNAPSHOT` to the Maven central snapshot repo. Please try it out.

**Comment by MieszkoMakuch** [comment]
I've checked that the fix in `0.6.0-SNAPSHOT` works as expected

**Comment by ennerf** [comment]
Thanks for confirming. It's released in version 0.5.10

**Comment by MieszkoMakuch** [comment]
Hi! I wanted to reopen this issue as we had cases where we cannot change the file format [as you suggested ](https://github.com/HebiRobotics/MFL/issues/69#issuecomment-1200286385) but we need to filter entries below the root level and selected cells in arrays of cells. 

I came up with a solution that adds custom Filter Api to Mat5Reader to allow this kind of filtering. I added a [preview pull request with changes](https://github.com/HebiRobotics/MFL/pull/72). It solves our case, however, it requires us to shadow the original Mat5Reader class with our modified class. 

I was wondering if you would consider adding such custom filtering to your library. Or maybe you could modify Mat5Reader class so that it can be extended without the need to shadow the entire class?

Thanks!

**Comment by ennerf** [comment]
Thanks for the PR. I'm not a fan of adding these filters directly to the Reader, but I think it's very reasonable to add an extension mechanism to enable your use case.

Would you mind creating another PR that exposes the fields you need to access to implement your filters as a subclass?

**Comment by MieszkoMakuch** [comment]
Sure. Soon I will add another PR with fields/methods that would be exposed.

**Comment by ennerf** [comment]
I left your changes as is and released it as `0.5.13`. Thanks.
